x/bank: doc & spec updates (#8999)
cosmos/cosmos-sdk: x/bank/spec/02_keepers.md

# Keepers

The bank module provides these exported keeper interfaces that can be
passed to other modules that read or update account balances. Modules
should use the least-permissive interface that provides the functionality they
require.

Best practices dictate careful review of `bank` module code to ensure that
permissions are limited in the way that you expect.

## Blacklisting Addresses

The `x/bank` module accepts a map of addresses that are considered blocklisted
from directly and explicitly receiving funds through means such as `MsgSend` and
`MsgMultiSend` and direct API calls like `SendCoinsFromModuleToAccount`.

Typically, these addresses are module accounts. If these addresses receive funds
outside of the expected rules of the state machine, invariants are likely to be

> broken and could result in a halted network.

Reviewer: Do we have any other example of account being blacklisted, which is not a module account?

Author: Nope :)

# Keepers

> The bank module provides three different exported keeper interfaces which can be
> passed to other modules which need to read or update account balances. Modules
> should use the least-permissive interface which provides the functionality they
> require.

Reviewer 2: ```suggestion
The bank module provides these exported keeper interfaces that can be
passed to other modules that read or update account balances. Modules
should use the least-permissive interface that provides the functionality they
require.
```

Reviewer 2: now we have more than three, best practice is to not count them "three exported keeper interfaces" becomes "these..."

> Note that you should always review the `bank` module code to ensure that
> permissions are limited in the way that you expect.

Reviewer 2: ```suggestion
Best practices dictate careful review of `bank` module code to ensure that
permissions are limited in the way that you expect.
```

## Blacklisting Addresses

> The `x/bank` module accepts a map of addresses that are considered blacklisted
> from directly and explicitly receiving funds through means such as `MsgSend` and
> `MsgMultiSend` in addition to direct API calls such as `SendCoinsFromModuleToAccount`.

Reviewer 2: ```suggestion
The `x/bank` module accepts a map of addresses that are considered blocklisted
from directly and explicitly receiving funds through means such as `MsgSend` and
`MsgMultiSend` and direct API calls like `SendCoinsFromModuleToAccount`.
```

> Typically, these addresses are module accounts and if they were to receive funds
> outside of the expected rules of the state-machine, invariants are likely to be
> broken and could result in a halted network.
> 
> By providing the `x/bank` module with a blacklisted set of addresses, if a user
> or client attempts to directly or indirectly, e.g. via IBC, send funds to any one
> of these blacklisted accounts, the operation will error.

Reviewer 2: ```suggestion
Typically, these addresses are module accounts. If these addresses receive funds
outside of the expected rules of the state machine, invariants are likely to be
broken and could result in a halted network.

By providing the `x/bank` module with a blocklisted set of addresses, an error occurs for the operation if a user or client attempts to directly or indirectly send funds to a blocklisted account, for example, by using [IBC](http://docs.cosmos.network/master/ibc/).
```

## SendKeeper

> The send keeper provides access to account balances and the ability to transfer coins between
> accounts, but not to alter the total supply (mint or burn coins).

Reviewer 2: ```suggestion
The send keeper provides access to account balances and the ability to transfer coins between
accounts. The send keeper does not alter the total supply (mint or burn coins).
```

## ViewKeeper

> The view keeper provides read-only access to account balances but no balance alteration
> functionality. All balance lookups are `O(1)`.

Reviewer 2: ```suggestion
The view keeper provides read-only access to account balances. The view keeper does not have balance alteration functionality. All balance lookups are `O(1)`.
```
